datafusion-examples/examples/default_column_values.rs

```diff
@@ -0,0 +1,366 @@
+#[derive(Debug)]
+struct DefaultValueTableProvider {
+    schema: SchemaRef,
+}
+
+impl DefaultValueTableProvider {
+    fn new(schema: SchemaRef) -> Self {
+        Self { schema }
+    }
+}
+
+#[async_trait]
+impl TableProvider for DefaultValueTableProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> Result<Vec<TableProviderFilterPushDown>> {
+        Ok(vec![TableProviderFilterPushDown::Inexact; filters.len()])
+    }
+
+    async fn scan(
+        &self,
+        state: &dyn Session,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        let schema = self.schema.clone();
+        let df_schema = DFSchema::try_from(schema.clone())?;
+        let filter = state.create_physical_expr(
+            conjunction(filters.iter().cloned()).unwrap_or_else(|| lit(true)),
+            &df_schema,
+        )?;
+
+        let parquet_source = ParquetSource::default()
+            .with_predicate(filter)
+            .with_pushdown_filters(true);
+
+        let object_store_url = ObjectStoreUrl::parse("memory://")?;
+        let store = state.runtime_env().object_store(object_store_url)?;
+
+        let mut files = vec![];
+        let mut listing = store.list(None);
+        while let Some(file) = listing.next().await {
+            if let Ok(file) = file {
+                files.push(file);
+            }
+        }
+
+        let file_group = files
+            .iter()
+            .map(|file| PartitionedFile::new(file.location.clone(), file.size))
+            .collect();
+
+        let file_scan_config = FileScanConfigBuilder::new(
+            ObjectStoreUrl::parse("memory://")?,
+            self.schema.clone(),
+            Arc::new(parquet_source),
+        )
+        .with_projection(projection.cloned())
+        .with_limit(limit)
+        .with_file_group(file_group)
+        .with_expr_adapter(Arc::new(DefaultValuePhysicalExprAdapter {
+            default_adapter: DefaultPhysicalExprAdapter,
+        }) as _);
+
+        Ok(Arc::new(DataSourceExec::new(Arc::new(
+            file_scan_config.build(),
+        ))))
+    }
+}
+
+/// Custom PhysicalExprAdapter that handles missing columns with default values from metadata
+/// and wraps DefaultPhysicalExprAdapter for standard schema adaptation
+#[derive(Debug)]
+struct DefaultValuePhysicalExprAdapter {
+    default_adapter: DefaultPhysicalExprAdapter,
+}
+
+impl PhysicalExprAdapter for DefaultValuePhysicalExprAdapter {
+    fn rewrite_to_file_schema(
+        &self,
+        expr: Arc<dyn PhysicalExpr>,
+        logical_file_schema: &Schema,
+        physical_file_schema: &Schema,
+        partition_values: &[(FieldRef, ScalarValue)],
```

Comment:
Not for this PR, but it seem to me that partition_values shouldn't really be in `PhysicalExprAdapter` -- rather the ListingTable should provide a PhysicalExprAdapter that knows how to fill in partition values

As a follow on PR perhaps

Reply:
Very interesting idea. I think let's see how the refactoring of `FileScanConfigBuilder` & co goes, then we can revisit moving the injection of the partition values around.

Reply:
Maybe we end up with `PhysicalExprAdapterFactory` that users pass in and then gets called as `let adapter = physical_expr_adapter_factory.create_adapter(logical_file_schema, physical_file_schema).with_partition_values(...);` and `let expr = adapter.rewrite(expr);`